<h1>Исследование надёжности заёмщиков</h1>

<h2>Шаг 1. Изучим информацию о данных.</h2>

In [1]:
import pandas as pd
df = pd.read_csv('/datasets/data.csv')

<h3>Посмотрим первые 5 строк</h3>

In [2]:
df.head()

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

<h3>Выведем всю информацию о DataFrame и посмотрим на типы данных</h3>

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


<h3>Проверим наличие пропусков</h3>

In [4]:
df.isnull().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

<h3>Проверим наличие дубликатов</h3>

In [5]:
df.duplicated().sum()

54

<h3>Пробежимся по столбцам и поищем аномалии</h3>

In [6]:
df['children'].value_counts() #считаю количество уникальных значений в столбце "количество детей в семье"

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

In [7]:
df['dob_years'].value_counts() #считаю количество уникальных значений в столбце "возраст клиента в годах"

35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
66    183
22    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

In [8]:
df['education'].value_counts() #считаю количество уникальных значений в столбце "уровень образования клиента"

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64

In [9]:
df['family_status'].value_counts() #считаю количество уникальных значений в столбце "семейное положение"

женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

In [10]:
df['gender'].value_counts() #считаю количество уникальных значений в столбце "пол клиента"

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

In [11]:
df['income_type'].value_counts() #считаю количество уникальных значений в столбце "тип занятости"

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
безработный            2
предприниматель        2
в декрете              1
студент                1
Name: income_type, dtype: int64

In [12]:
df['debt'].value_counts() #считаю количество уникальных значений в столбце "имел ли задолженность по возврату кредитов"

0    19784
1     1741
Name: debt, dtype: int64

In [13]:
df['purpose'].value_counts() #считаю количество уникальных значений в столбце "цель получения кредита"

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
покупка жилья для сдачи                   653
операции с жильем                         653
операции с коммерческой недвижимостью     651
жилье                                     647
покупка жилья                             647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
строительство недвижимости                620
покупка своего жилья                      620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

<p><span style="font-size: 20px; font-weight: bold;">Вывод:</span><br><br> Что мы имеем? Смотря на таблицу, первое что бросается в глаза - это отрицательные значения в столбце <em>days_employed</em> и <em>children</em>. В столбце <em>children</em> 76 человек имеет по 20 детей o_O также, пропуски имеются в двух столбцах <em>days_employed</em> и <em>total_income</em>. Всего их 2174 что ~ 10%. Почему так? Общий трудовой стаж в днях и ежемесячный доход взаимосвязаны. <span style="color: green;">Нет стажа - нет дохода</span>, все логично. Но это может быть обьяснено тем, что некоторые люди <span style="color: red">работают неофициально</span>, некоторые <span style="color: red">безработные студенты</span> и тд. По всей таблице имеем 54 дубликата. В столбце <em>education</em> имеем проблему со значениями, <span style="color: red;">регистром.</span> И последнее, в столбце <em>purpose</em> имеем 38 уникальных значений, их необходимо будет лемматизировать</p>

<h2>Шаг 2. Предобработаем данные.</h2>

<h3>Заменяем отрицательные значения в столбце <em>days_employed</em></h3>

In [14]:
df['days_employed'] = abs(df['days_employed']) #использую готовую функцию для конвертации в положительное число
df['days_employed'].head()

0      8437.673028
1      4024.803754
2      5623.422610
3      4124.747207
4    340266.072047
Name: days_employed, dtype: float64

<h3>Удаляем аномальные значения в столбце <em>children</em></h3>

In [15]:
df = df[(df['children'] >= 0) & (df['children'] <= 5)]
df.sort_values('children')['children'].value_counts().sort_index()

0    14149
1     4818
2     2055
3      330
4       41
5        9
Name: children, dtype: int64

<h3>Заменяем пропущенные значения в столбцах <em>days_employed</em> и <em>total_income</em></h3>

In [16]:
days_employed_median = df['days_employed'].median() #считаю медиану столбца "days_employed" 2194.22
total_income_median = df['total_income'].median() #считаю медиану столбца "total_income" 145017.93
df['days_employed']= df['days_employed'].fillna(value=days_employed_median) #заменяю пустые значения на медиану столбца "days_employed"
df['total_income']= df['total_income'].fillna(value=total_income_median) #заменяю пустые значения на медиану столбца "total_income"

In [17]:
df.isnull().sum() #после замены, пустых значений нет

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

<h3>Переведем значения столбца <em>days_employed</em> в целочисленный тип</h3>

In [18]:
df['days_employed'] = df['days_employed'].astype('int')

In [19]:
df.head()

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   Среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

<p><span style="font-size: 20px; font-weight: bold;">Вывод:</span><br><br><ul><li>Заменили отрицательные значения в столбце <em>days_employed</em> при помощи метода astype</li><li>Заменили пропущенные значения в столбцах <em>days_employed</em> и <em>total_income</em></li><li>Переведем значения столбца <em>days_employed</em> в целочисленный тип</li><li>Также выяснили, медианный стаж работы = 2194.22 - 6 лет, а медианный ежемесячный доход = 155017.93</li><li>Избавились от записей, в которых есть ошибочные данные в столбце <em>children</em></li></ul></p>
<p style="color: green;">Избавились от пропусков, оставив только необходимые для будущего анализа данные</p>

<h3>Обработаем дубликаты</h3>

In [20]:
df['education'] = df['education'].str.lower() #привожу данные к нижнему регистру  

In [21]:
df.sort_values('education')['education'].value_counts()

среднее                15136
высшее                  5237
неоконченное высшее      741
начальное                282
ученая степень             6
Name: education, dtype: int64

<h3>Лемматизируем</h3>

In [22]:
from collections import Counter

from pymystem3 import Mystem
m = Mystem()

# Находим самые частовстречающиеся слова в леммах
purposes = []

for element in df['purpose']:
    lemma = m.lemmatize(element)
    purposes +=  lemma

print(Counter(purposes))

Counter({' ': 33485, '\n': 21402, 'недвижимость': 6330, 'покупка': 5880, 'жилье': 4450, 'автомобиль': 4288, 'образование': 3997, 'с': 2906, 'операция': 2593, 'свадьба': 2337, 'свой': 2224, 'на': 2218, 'строительство': 1870, 'высокий': 1368, 'получение': 1311, 'коммерческий': 1307, 'для': 1291, 'жилой': 1225, 'сделка': 939, 'заниматься': 908, 'дополнительный': 902, 'проведение': 772, 'сыграть': 769, 'сдача': 651, 'семья': 640, 'собственный': 628, 'со': 626, 'ремонт': 609, 'подержанный': 484, 'подержать': 472, 'приобретение': 460, 'профильный': 432})


<h3>Выделяем самые частовстречающиеся слова</h3>

In [23]:
purpose_category = [
    'недвижимость',
    'жилье',
    'автомобиль',
    'образование',
    'свадьба'
]

<h3>Опишем функцию добавления нового столбца</h3>

In [24]:
def categorize_purpose(text): #Функция возвращает краткую цель кредита на основе лемматизированного текста. Со столбцом будет удобнее работать.
    lemmas = m.lemmatize(text)
    for lemma in lemmas:
        for category in purpose_category:
            if category in lemma:
                return category

In [25]:
#добавляем новый столбец с целями кредита, используя метод apply()
df['purpose_category'] = df['purpose'].apply(categorize_purpose)

In [26]:
#выведем таблицу
df.head(10)

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   
5         0            926         27    высшее             0   
6         0           2879         43    высшее             0   
7         0            152         50   среднее             1   
8         2           6929         35    высшее             0   
9         0           2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose purpose_category  
0               покупка жилья            жилье  
1     приобретение автомобиля       автомобиль  
2               покупка жилья            жилье  
3  дополнительное образование      образование  
4             сыграть свадьбу          свадьба  
5               покупка жилья            жилье  
6           операции с жильем            жилье  
7                 образование      образование  
8       на проведение свадьбы          свадьба  
9     покупка жилья для семьи            жилье

<p><span style="font-size: 20px; font-weight: bold;">Вывод:</span><br><br>Из большинства значения в столбце "цель получения кредита" мы смогли выявить и категоризировать <span style="color: green;">всего 5 значений.</span> Теперь намного удобнее работать с таблицей.</p>

<h3>Категоризация данных</h3>

<p>Чтобы ответить на вопросы, необходимо категоризировать пользователей.</p>

<p>1. По количеству детей:</p>
<ul>
    <li>нет детей - <span style="color: green;">бездетная</span></li>
    <li>1 ребенок - <span style="color: green;">однодетная</span></li>
    <li>2 ребенка - <span style="color: green;">малодетная</span></li>
    <li>3 и больше - <span style="color: green;">многодетная</span></li>
</ul>

In [27]:
#напишем для этого функцию котороя по количеству возвращает ответ
def categoriez_children(count):
    if count == 0:
        return 'бездетная'
    elif count == 1:
        return 'однодетная'
    elif count == 2:
        return 'малодетная'
    else:
        return 'многодетная'

In [28]:
df['children_category'] = df['children'].apply(categoriez_children)

In [29]:
df.head(5)

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose purpose_category children_category  
0               покупка жилья            жилье        однодетная  
1     приобретение автомобиля       автомобиль        однодетная  
2               покупка жилья            жилье         бездетная  
3  дополнительное образование      образование       многодетная  
4             сыграть свадьбу          свадьба         бездетная

<p>2. По семейному положению. Всего у нас 5 значений, можно сузить:</p>
<ul>
    <li>женат / замужем - <span style="color: green;">В браке</span></li>
    <li>гражданский брак, в разводе, вдовец / вдова  - <span style="color: green;">Не в браке</span></li>
</ul>

In [30]:
df['family_status'].value_counts()

женат / замужем          12302
гражданский брак          4160
Не женат / не замужем     2799
в разводе                 1189
вдовец / вдова             952
Name: family_status, dtype: int64

In [31]:
#напишем для этого функцию котороя по передаваемому значению возвращает ответ
def categoriez_family_status(value):
    if value == 'женат / замужем':
        return 'в браке'
    else:
        return 'не в браке'

In [32]:
df['family_status_category'] = df['family_status'].apply(categoriez_family_status)

In [33]:
df.head()

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose purpose_category children_category  \
0               покупка жилья            жилье        однодетная   
1     приобретение автомобиля       автомобиль        однодетная   
2               покупка жилья            жилье         бездетная   
3  дополнительное образование      образование       многодетная   
4             сыграть свадьбу          свадьба         бездетная   

  family_status_category  
0                в браке  
1                в браке  
2                в браке  
3                в браке  
4             не в браке

<p>3. По доходу.</p>
<ul>
    <li> <= 20667 - <span style="color: green;">низкий доход</span></li>
    <li> <= 145020 - <span style="color: green;">средний доход</span></li>
    <li> > - <span style="color: green;">высокий доход</span></li>
</ul>

In [44]:
min_total_income = df['total_income'].min()
min_total_income

20667.26379327158

In [35]:
df['total_income'].max()

2265604.028722744

In [43]:
median_total_income = df['total_income'].median()
median_total_income

145020.80127962783

In [37]:
def categoriez_total_income(value):
    if value <= min_total_income:
        return 'низкий доход'
    elif value <= median_total_income:
        return 'средний доход'
    else:
        return 'высокий доход'

In [38]:
df['total_income_category'] = df['total_income'].apply(categoriez_total_income)

In [39]:
df.head()

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose purpose_category children_category  \
0               покупка жилья            жилье        однодетная   
1     приобретение автомобиля       автомобиль        однодетная   
2               покупка жилья            жилье         бездетная   
3  дополнительное образование      образование       многодетная   
4             сыграть свадьбу          свадьба         бездетная   

  family_status_category total_income_category  
0                в браке         высокий доход  
1                в браке         средний доход  
2                в браке         высокий доход  
3                в браке         высокий доход  
4             не в браке         высокий доход

<p><span style="font-size: 20px; font-weight: bold;">Вывод:</span><br><br>Мы категоризовали данные, далее группируем их и отвечаем на вопросы.</p>

<h2>Шаг 3. Отвечаем на вопросы.</h2>

<h3>1. Есть ли зависимость между наличием детей и возвратом кредита в срок?</h3>

In [40]:
pd.pivot_table(df, index='children_category', values='debt')

debt
children_category          
бездетная          0.075129
малодетная         0.094404
многодетная        0.081579
однодетная         0.092154

<p><span style="font-size: 20px; font-weight: bold;">Вывод:</span><br><br>Исходя из группировки, видно, что зависимости от количества детей и возвратом денег нет.</p><p>Бездетные семьи (ожидаемо) чаще выплачивают кредит.</p>

<h3>2. Есть ли зависимость между семейным положением и возвратом кредита в срок?</h3>

In [41]:
pd.pivot_table(df, index='family_status_category', values='debt')

debt
family_status_category          
в браке                 0.075354
не в браке              0.088462

<p><span style="font-size: 20px; font-weight: bold;">Вывод:</span><br><br>Исходя из группировки, видно, что клиенты, которые находятся в браке реже остаются в должниках. Возможно это потому, что не один человек работае и оплачивает кредит, а работают и оплачивают они уже вдвоем (плюсом может помогают дети).</p>

<h3>3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</h3>

In [42]:
pd.pivot_table(df, index='total_income_category', values='debt')

debt
total_income_category          
высокий доход          0.079210
низкий доход           1.000000
средний доход          0.082251

<p><span style="font-size: 20px; font-weight: bold;">Вывод:</span><br><br>Оплата кредита напрямую зависит от дохода. Чем выше доход - тем меньше процентов задолженности.</p>

<h3>4.  разные цели кредита влияют на его возврат в срок?</h3>

In [47]:
pd.pivot_table(df, index='purpose_category', values='debt')

debt
purpose_category          
автомобиль        0.093284
жилье             0.069213
недвижимость      0.074566
образование       0.092319
свадьба           0.078306

<p><span style="font-size: 20px; font-weight: bold;">Вывод:</span><br><br>Исходя из группировки видно, что задолженности по кредитам за автомобиль и образование имеют самый большой процент. Интересно, что за жилье - самый маленький.</p>

<h2>Шаг 4. Общий вывод.</h2>

<p>Исходя из проделанной работы, мы можем сделать вывод: Кому лучше всего давать кредит?</p>
<ul>
    <li style="color: green;">Тем, кто берет на покупку жилья, недвижимости или свадбу</li>
    <li style="color: green;">Тем, у кого нет детей</li>
    <li style="color: green;">Тем, кто находится в браке</li>
    <li style="color: green;">Тем, кто имеет высокий доход</li>
</ul>